### PR TITLE
Separate out predicates from ty::Generics and ty::TraitDef

### PR DESCRIPTION
**Pull request by nikomatsakis**
This resolves a number of bugs that trigger stack overflows or other cyclic errors (#20220, #20551, at least). I plan to do some follow-up PRs addressing further problems and also simplifying / restructuring collect, but I wanted to get something landed for now to unblock #20551 in particular.

r? @nick29581 (it is based on work that you started)
f? @jroesch (also based on your branch)

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Oh, shoot, tidy failures. I'll patch those.

File: src/librustc/util/ppaux.rs [hunk 1]
```diff
@@ -480,6 +497,13 @@ pub fn parameterized<'tcx>(cx: &ctxt<'tcx>,
         }
     }
 
+    // It is important to execute this conditionally, only if -Z
+    // verbose is false. Otherwise, debug logs can sometimes cause
+    // ICEs trying to fetch the generics early in the pipeline. This
+    // is kind of a hacky workaround in that -Z verbose is required to
+    // avoid those ICEs.
+    let generics = get_generics();
```
Comment (jroesch) [inline src/librustc/util/ppaux.rs]: :+1: it is better than an ICE and having to remove all the debugging :)

**Comment by jroesch** [comment]
I read it over and it looks good. I'm assuming you are waiting to touch up and land the rest of the changes from my branch? I'm sure there issues to be ironed out with it still.

**Comment by nikomatsakis** [comment]
@jroesch yes once this lands I plan to rebase your branch on top of it (as well as some other unrelated cleanup I'd like to do).

File: src/librustc/metadata/csearch.rs [hunk 2]
```diff
@@ -233,6 +233,14 @@ pub fn get_trait_def<'tcx>(tcx: &ty::ctxt<'tcx>, def: ast::DefId) -> ty::TraitDe
     decoder::get_trait_def(&*cdata, def.node, tcx)
 }
 
+pub fn get_predicates<'tcx>(tcx: &ty::ctxt<'tcx>, def: ast::DefId)
+                                  -> ty::GenericPredicates<'tcx>
```
Comment (nick29581) [inline src/librustc/metadata/csearch.rs]: indent

File: src/librustc/metadata/decoder.rs [hunk 3]
```diff
@@ -417,16 +417,22 @@ pub fn get_trait_def<'tcx>(cdata: Cmd,
...
 }
 
-pub fn get_type<'tcx>(cdata: Cmd, id: ast::NodeId, tcx: &ty::ctxt<'tcx>)
-    -> ty::TypeScheme<'tcx> {
-
-    let item = lookup_item(id, cdata.data());
+pub fn get_predicates<'tcx>(cdata: Cmd,
+                                  item_id: ast::NodeId,
+                                  tcx: &ty::ctxt<'tcx>)
+                                  -> ty::GenericPredicates<'tcx>
```
Comment (nick29581) [inline src/librustc/metadata/decoder.rs]: indents (and you thought make tidy was nit-picky)

**Comment by nikomatsakis** [comment]
Indentation nits addressed.

File: src/librustc_typeck/collect.rs [hunk 4]
```diff
@@ -20,28 +20,83 @@ current crate: we assume that after the collect phase, the types of
...
+2. `Foo<X>` where `Foo` is a type alias requires knowing what `Foo` expands to
+3. Translating `SomeTrait` with no explicit lifetime bound requires knowing
+   - supertraits of `SomeTrait`
+4. Translating `T::X` (vs `<T as Trait>::X`) requires knowing
+   - bounds on `T`
+   - supertraits of those bounds
+
+So as you can see, in general translating types requires knowing the
+trait hierarchy. But this gets a bit tricky because translating the
+trait hierarchy requires convering the types that appear in trait
```
Comment (nick29581) [inline src/librustc_typeck/collect.rs]: typo 'convering'

File: src/librustc_typeck/collect.rs [hunk 5]
```diff
@@ -1573,27 +1765,28 @@ fn check_method_self_type<'a, 'tcx, RS:RegionScope>(
...
-        let infcx = infer::new_infer_ctxt(ccx.tcx);
-        drop(::require_same_types(ccx.tcx,
+                    tcx, body_scope, &ty::Binder(base_type)));
+
+        //debug!("required_type={} required_type_free={} \
+                //base_type={} base_type_free={}",
+               //required_type.repr(tcx),
+               //required_type_free.repr(tcx),
+               //base_type.repr(tcx),
+               //base_type_free.repr(tcx));
```
Comment (nick29581) [inline src/librustc_typeck/collect.rs]: should be removed or uncommented

**Comment by nikomatsakis** [comment]
Rebased, nits addressed.

**Pull request merged**